### PR TITLE
Refactor scheduler interface FitPredicate

### PR DESCRIPTION
**Pull request by anfernee**
**What this PR does / why we need it**: Remove duplicates that handles multiple predicate. Cleaner interface.

**Special notes for your reviewer**:

**Release note**:
```release-note
None
```

- Rename original FitPredicate to FitPredicateFunction because it's a fuction
- Make FitPredicate an interface which has a Predicate method
- Make FitPredicateFunction implement FitPredicate
- Add FitPredicates to handle a group of FitPredicates
- Make FitPredicates implement FitPredicate

### CODE REVIEW

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 1]
```diff
@@ -594,7 +594,7 @@ func podName(pod *v1.Pod) string {
 // PodFitsResources checks if a node has sufficient resources, such as cpu, memory, gpu, opaque int resources etc to run a pod.
 // First return value indicates whether a node has sufficient resources to run a pod while the second return value indicates the
 // predicate failure reasons if the node has insufficient resources to run the pod.
-func PodFitsResources(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+var PodFitsResources algorithm.FitPredicateFunction = func(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I don't like this change. Don't change like this.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -716,7 +716,7 @@ func podMatchesNodeLabels(pod *v1.Pod, node *v1.Node) bool {
 }
 
 // PodMatchNodeSelector checks if a pod node selector matches the node label.
-func PodMatchNodeSelector(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+var PodMatchNodeSelector algorithm.FitPredicateFunction = func(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Ditto

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 3]
```diff
@@ -728,7 +728,7 @@ func PodMatchNodeSelector(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInf
 }
 
 // PodFitsHost checks if a pod spec node name matches the current node.
-func PodFitsHost(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
+var PodFitsHost algorithm.FitPredicateFunction = func(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *schedulercache.NodeInfo) (bool, []algorithm.PredicateFailureReason, error) {
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Ditto

**Review by dixudx — CHANGES_REQUESTED** [review]

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 1 above)
Comment (anfernee) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I can keep it as before, but every time you use it you have to use it like this: `algorithm.FitPredicateFunction(PodFitsResources)`.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 4]
```diff
@@ -930,71 +930,22 @@ func haveSame(a1, a2 []string) bool {
...
-	if err != nil {
-		return false, predicateFails, err
-	}
-	if !fit {
-		predicateFails = append(predicateFails, reasons...)
-	}
-	return len(predicateFails) == 0, predicateFails, nil
-}
+var EssentialPredicates = algorithm.FitPredicates([]algorithm.FitPredicate{
+	PodFitsHost,
```
Comment (anfernee) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: for instance, this: `algorithm.FitPredicateFunction(PodFitsHost)`. However, `PodFitsHost` is a predicate per se, it's not a function.

**Comment by anfernee** [comment]
I think the failure is irrelevant. @wojtek-t @jayunit100 can you guys take a look? thanks.

**Comment by jayunit100** [comment]
I think the change might pave the way for more composability for external schedulers in the future - but otherwise am +0 given that it's debatable that this introduced interface has more then one method that we need .... maybe let's punt this to folks like quinton Hoole or others that are working on new scheduler implementations for feedback .

**Comment by anfernee** [comment]
thanks @jayunit100 I am not aware that there are other folks are actually working on a new scheduler implementation. It's interesting. I have some idea of changing priority functions too, which I believe could dedup current code base, and make the logic clearer. anyway, let me ping Quinton Hoole to see if I could help. Thanks again.

**Comment by anfernee** [comment]
I know the interface change would affect external schedulers, but i still feel it's the right thing to do. but i am open to keep the old ones. maybe it should be noted in release note.

**Comment by anfernee** [comment]
related to #24246

**Comment by anfernee** [comment]
cc @kubernetes/sig-scheduling-pr-reviews

**Comment by anfernee** [comment]
@quinton-hoole do you have concerns?

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 1 above)
Comment (wojtek-t) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: +1 to what @dixudx wrote, I don't like changing functions to vars.

**Comment by bsalamat** [comment]
As others mentioned, I feel converting functions to variables reduces readability. This is a pattern used in many places in this PR. So, I am not sure if this PR improves readability.

**Pull request closed without merging**